controllers/apps/operations/ops_manager.go

```diff
@@ -231,6 +231,37 @@ func (opsMgr *OpsManager) validateDependOnSuccessfulOps(reqCtx intctrlutil.Reque
 	return true, nil
 }
 
+// handleOpsIsRunningTimedOut handles if the opsRequest is timed out.
+func (opsMgr *OpsManager) handleOpsIsRunningTimedOut(reqCtx intctrlutil.RequestCtx,
+	cli client.Client,
+	opsRes *OpsResource,
+	requeueAfter time.Duration) (time.Duration, error) {
+	timeoutSeconds := opsRes.OpsRequest.Spec.TimeoutSeconds
+	if timeoutSeconds == nil || *timeoutSeconds == 0 {
+		return requeueAfter, nil
+	}
+	checkTimedOut := func(startTime metav1.Time) (time.Duration, error) {
+		timeoutPoint := startTime.Add(time.Duration(*timeoutSeconds))
+		if !time.Now().Before(timeoutPoint) {
+			return 0, PatchOpsStatus(reqCtx.Ctx, cli, opsRes, appsv1alpha1.OpsAbortedPhase,
+
+				appsv1alpha1.NewAbortedCondition("Aborted due to exceeding the specified timeout period (timeoutSeconds)"))
+		}
+		if requeueAfter != 0 {
+			return requeueAfter, nil
+		}
+		return time.Until(timeoutPoint), nil
+	}
+	if !opsRes.OpsRequest.Spec.Cancel {
+		return checkTimedOut(opsRes.OpsRequest.Status.StartTimestamp)
+	}
+	if opsRes.OpsRequest.Status.CancelTimestamp.IsZero() {
+		return requeueAfter, nil
+	}
+	// After canceling opsRequest, recalculate whether the opsRequest runs out of time based on the cancellation time point
+	return checkTimedOut(opsRes.OpsRequest.Status.CancelTimestamp)
```

Comment:
This behavior may seem counterintuitive. When an opsRequest is canceled, it might actually run longer than if it were not canceled. This is because the timeout is recalculated based on the cancellation timestamp rather than the original start timestamp.